ProcessList.c

```diff
@@ -114,31 +114,66 @@ int ProcessList_size(ProcessList* this) {
    return (Vector_size(this->processes));
 }
 
+// ProcessList_updateTreeSetLayer sort this->treeSet,
+// relaying only on itself.
+//
+// 'newTree' is a temporaraly Hashtable which at the end of the function
+// will hold a renovated 'treeSet'.
+// It's necessary to use a temporaraly Hashtable as it's not save to traverse this->treeSet
+// and at the same time make changes in it.
+//
+// It could though be removed with sort by TREE_SET if we could savely update this->processes in place.
+//
+// left_bound and right_bound mainly used as a optimization to cut the list where we are building a 'layer'.
 static void ProcessList_updateTreeSetLayer(ProcessList* this, Hashtable* newTree, unsigned int left_bound, unsigned int right_bound, unsigned int deep, unsigned int left, unsigned int right, unsigned int* index, unsigned int* tree_index, int indent) {
-   // it's guaranted that layer_size is enough space
-   // but most likely it needs less
+   // It's guaranted that layer_size is enough space
+   // but most likely it needs less. Specificaly on first iteration.
    int layer_size = (right - left) / 2;
    Vector* layer = Vector_new(this->processes->type, false, layer_size);
 
+   // Find all process on the same layer (process with the same deep value
+   // and included in a range from left_bound to right_bound.
+   //
+   // Also this loops keeps track of left_bound and right_bound of these proceses to not lose this information
+   // after the list is sorted.
+   //
+   // left_bound and right_bound doesn't represent the same what lhs and rhs represents.
+   // Here's an example why. The nested tree set sorted by left value. Which guaranted have been called when
+   // this function is called.  
+   // 
+   // i | l | r
+   // 1 | 1 | 9 
+   // 2 | 2 | 8
+   // 3 | 4 | 5
+   // 4 | 6 | 7
```

Comment:
The explanation for the example is not very intuitive.

Let me rephrase a bit, maybe it's easier to digest with another wording:

```
The variables left_bound and right_bound are different from what the values lhs and rhs represent,
While left_bound and right_bound define a range of processes to look at, the values given by lhs and rhs are indices into an array …

In the below example note how filtering a range of indices i is different from filtering for processes in the bounds left_bound < x < right_bound …

The nested tree set is sorted by left value, which is guaranteed upon entry/exit(???) of this function.
```

Maybe something more along those lines? I hope I got it somewhat correct. Also the naming of things in the comment should follow the variables used in the code to avoid confusion when trying to understand the comment.

Reply:
> The nested tree set is sorted by left value, which is guaranteed upon entry/exit(???) of this function.

Yes

So as I said in the doc comment for function if there's a save way to sort `ProcessList::processes` in place we could remove newTreeSet and a sort at the end of update algorithm

Reply:
I could you've got more versed in these things so if you could `suggest` your variant of a comment I could commit it 😄

Reply:
The only in-place variant to sort I can think of right now would cause sorting to be O(n²) or worse …

To avoid having to update tree_left/tree_right for each process (and its tree children) for each move, just track how much each parent is moved starting from a first, consistent state to reach its final place. In a final run over all processes sum up the offsets for each process' parents and update tree_left/tree_right accordingly.

Given:
```
pid | ppid | depth |  left | right | move | xpnd
------------------------------------------------
  1 |    0 |     0 |     1 |     2 |    0 |    0
  2 |    1 |     0 |     3 |     4 |    0 |    0
  3 |    2 |     0 |     5 |     6 |    0 |    0
  4 |    1 |     0 |     7 |     8 |    0 |    0
  5 |    2 |     0 |     9 |    10 |    0 |    0
```

Wanted:
```
pid | ppid | depth |  left | right | move | xpnd
------------------------------------------------
  1 |    0 |     0 |     1 |    10 |    0 |    4
  2 |    1 |     1 |     2 |     7 |   -1 |    4
  3 |    2 |     2 |     3 |     4 |   -1 |    0
  4 |    1 |     1 |     8 |     9 |   -3 |    0
  5 |    2 |     2 |     5 |     6 |   -3 |    0
```

move: distance moved relative to parent, absolute move distance results from adding move value including all parent's.
xpnd: change of distance between left and right. total change by adding xpnd from all children.

Moving one process at a time to the proper place yields said O(n²) behaviour. You can do better, if you can predict the correct insert location for each process by:
1. only inserting at the end of a child list
2. consolidating that sort order (eliminates xpnd variation, resets move)
3. sort sub-lists by determining moves (updates move only)
4. consolidate movements

Consolidation can always be done in linear time if you iterate the structure and have consistent move/xpnd values to work with. The complexity goes into preparing those values.

Wondering how much this already resembles the current PRs implementation.

Reply:
Aa maybe there's something in keeping track of processing by saving offset value.
But currently nothings like that there 😄 . Even though I think that current implementation is generally faster then O(n²) as we reduce an process array considerably by using `left_bound`, `right_stuff`.

Is it save to remove elements from `this->processes`? If it is sorting the array at the end could be deleted.